File: R/PANPRS_gsPEN.R
#' Run the gsPEN algorithm for multiple traits, without functional annotations.
#' @param summary_z A matrix of summary statistics for each SNP and trait.
#' @param n_vec A vector of sample sizes for each of the Q traits corresponding to the Q columns of summary_z.
#' @param plinkLD A matrix of LD values for each pair of SNPs.
#' @param n_iter The number of iterations to run the algorithm.
#' @param upper_val The upper bound for the tuning parameter.
#' @param breaking The number of iterations to run before checking for convergence.
#' @param z_scale The scaling factor for the summary statistics.
#' @param tuning_matrix A matrix of tuning parameters.
#' @param tau_factor A vector of factors to multiply the median value by to get the tuning parameters.
#' @param len_lim_lambda The number of tuning parameters to use for the first iteration.
#' @param sub_tuning The number of tuning parameters to use for the second iteration.
#' @param lim_lambda The range of tuning parameters to use for the first iteration.
#' @param len_lambda The number of tuning parameters to use for the second iteration.
#' @param df_max The maximum degrees of freedom for the model.
#' @param sparse_beta Whether to use the sparse version of the algorithm.
#' @param debug_output Whether to output the tuning combinations that did not converge.
#' @param verbose Whether to output information through the evaluation of the algorithm.
#' @return A named list containing the following elements:
#' beta_matrix: A matrix of the estimated coefficients for each SNP and trait.
#' num_iter_vec: A vector of the number of iterations for each tuning combination.
#' all_tuning_matrix: A matrix of the tuning parameters used for each tuning combination.
#' @importFrom stats median qnorm quantile
#' @examples
#' # Load the library and data
#' library(PANPRSnext)
#' data("summaryZ")
#' data("Nvec")
#' data("plinkLD")
#'
#' # Take random subset of the data
#' subset <- sample(nrow(summaryZ), 100)
#' subset_summary_z <- summaryZ[subset, ]
#'
#' # Run gsPEN
#' output <- gsPEN_R(
#'   summary_z = subset_summary_z,
#'   n_vec = Nvec,
#'   plinkLD = plinkLD
#' )
#' @export
gsPEN_R <- function(
    summary_z,
    n_vec,
    plinkLD,
    n_iter = 100,
    upper_val = NULL,
    breaking = 1,
    z_scale = 1,
    tuning_matrix = NULL,
    tau_factor = c(1 / 25, 1, 10),
    len_lim_lambda = 10,
    sub_tuning = 50,
    lim_lambda = c(0.5, 0.9),
    len_lambda = 200,
    df_max = NULL,
    sparse_beta = FALSE,
    debug_output = FALSE,
    verbose = FALSE) {
  if (z_scale != 1) {
    stop("Tuning values set-up for multiple traits analysis requires z_scale = 1") # nolint: object_usage_linter.
  }

  P <- nrow(summary_z)
  Q <- length(n_vec)

  summary_betas <- matrix(0, nrow = P, ncol = Q)
  SD_vec <- matrix(0, nrow = P, ncol = Q)

  for (i in 1:Q) {
    summary_betas[, i] <- summary_z[, i] / sqrt(n_vec[i])
    SD_vec[, i] <- 1 / sqrt(n_vec[i])
  }

  rownames(summary_betas) <- rownames(summary_z)

  if (is.null(df_max)) {
    df_max <- ceiling(0.7 * P)
  }

  if (is.null(tuning_matrix)) {
    median_val <- median(apply(abs(summary_betas), 1, sum), na.rm = TRUE)
    tau_vec <- sort(median_val * tau_factor)
    lim_lambda <- quantile(abs(summary_z[, 1]), lim_lambda)

    tuning_matrix <- Tuning_setup_group_only(
      tau_vec,
      sub_tuning,
      lim_lambda,
      len_lambda,
      median_val,
      Q
    )
  }

  beta_index <- c(seq_len(nrow(summary_betas))) - 1
  SNP_names <- rownames(summary_betas)

  # Takes only the SNPs that are present in both the PLINK data set and the GWAs
  ld_J <- PlinkLD_transform(plinkLD, SNP_names)

  J_index_matrix <- matrix(nrow = nrow(ld_J), ncol = 2)

  mat1 <- match(ld_J[, 1], SNP_names)
  mat2 <- match(ld_J[, 2], SNP_names)

  J_index_matrix[, 1] <- beta_index[mat1]
  J_index_matrix[, 2] <- beta_index[mat2]

  ld_J[, 1] <- J_index_matrix[, 1]
  ld_J[, 2] <- J_index_matrix[, 2]

  od <- order(J_index_matrix[, 1], J_index_matrix[, 2], decreasing = FALSE)
  ld_J <- ld_J[od, ]


  wind <- which(!beta_index %in% ld_J[, 1])
  num_indices <- length(wind)

  index_J <- -1
  if (num_indices > 0) {
    index_J <- beta_index[wind]
  }

  counts <- table(ld_J[, 1])
  num_SNP <- length(counts)

  index_S <- c(0, cumsum(counts)[-num_SNP])
  index_E <- cumsum(counts) - 1

  index_matrix <- matrix(nrow = num_SNP, ncol = 3)
  index_matrix[, 1] <- as.numeric(names(counts))

  index_matrix[, 2] <- index_S
  index_matrix[, 3] <- index_E

  ld_vec <- ld_J[, 3]
  ld_J <- ld_J[, 2]


  if (is.null(upper_val)) {
    upper_val <- ceiling(max(abs(summary_betas), na.rm = TRUE) * 50)
  }

  nrow_index_matrix <- nrow(index_matrix)
  ncol_index_matrix <- ncol(index_matrix)

  nrow_tuning_matrix <- nrow(tuning_matrix)
  ncol_tuning_matrix <- ncol(tuning_matrix)

  nrow_beta_matrix <- nrow_tuning_matrix
  ncol_beta_matrix <- P * Q

  dims <- c(
    num_SNP, # 1
    P, # 2
    Q, # 3
    nrow_index_matrix, # 4
    ncol_index_matrix, # 5
    nrow_tuning_matrix, # 6
    ncol_tuning_matrix, # 7
    nrow_beta_matrix, # 8
    ncol_beta_matrix # 9
  )

  params <- c(
    upper_val, # 1
    n_iter, # 2
    breaking, # 3
    z_scale, # 4
    df_max, # 5
    num_indices # 6
  )

  args <- list(
    summary_betas,
    ld_J,
    index_matrix,
    index_J,
    ld_vec,
    SD_vec,
    tuning_matrix,
    dims,
    params
  )

  if (verbose) {
    cat("Number of total tuning combinations =", nrow_tuning_matrix, "\n")
  }

  if (sparse_beta) {
    Z <- do.call(
      gsPEN_sparse_cpp,
      args
    )
  } else {
    Z <- do.call(
      gsPEN_cpp,
      args
    )
  }

  beta_matrix <- Z$beta_matrix
  colnames(beta_matrix) <- paste0(
    rep(SNP_names, times = Q),
    ".trait",
    rep(c(1:Q), each = P)
  )

  tuning_matrix <- Z$tuning_matrix
  colnames(tuning_matrix) <- c(
    "lambda1",
    "null",
    "lambda2",
    "tau"
  )
  tuning_matrix <- tuning_matrix[, c(1, 3, 4)]
  tuning_matrix[c(1:len_lambda), c(2:3)] <- NA

  num_iter_vec <- Z$num_iter_vec

  # Remove the tuning combinations that did not converge (correspons to -2 in num_iter_vec)
  if (!debug_output) {
    converge_index <- which(num_iter_vec > 0)
    if (verbose) {
      cat("Removing (", length(num_iter_vec) - length(converge_index), " ) tuning combinations that did not converge\n")
    }
    num_iter_vec <- num_iter_vec[converge_index]
    beta_matrix <- beta_matrix[converge_index, ]
    tuning_matrix <- tuning_matrix[converge_index, ]
  }

  output <- Clean_results(
    beta_matrix = beta_matrix,
    num_iter_vec = num_iter_vec,
    all_tuning_matrix = tuning_matrix
  )

  return(output)
}

File: src/PANPRS_gsPEN.cpp
// Prevent bounds checking for increased performance
#define ARMA_NO_DEBUG 1

#include <RcppArmadillo.h>

#include <stdio.h>
#include <stdlib.h>

#include "PANPRS_gsPEN.h"

Rcpp::List gsPEN_cpp(
    arma::Mat<double> summary_betas,
    arma::Col<int> ld_J,
    arma::Mat<int> index_matrix,
    arma::Col<int> index_J,
    arma::Col<double> ld_vec,
    arma::Mat<double> SD_vec,
    arma::Mat<double> tuning_matrix,
    arma::Col<int> dims,
    arma::Col<double> params)
{
  int num_SNP = dims(0);

  int P = dims(1);
  int Q = dims(2);

  int nrow_tuning_matrix = dims(5);

  int nrow_beta_matrix = dims(7);
  int ncol_beta_matrix = dims(8);

  double upper_val = params(0);
  int num_iter = params(1);
  int breaking = params(2);
  int z_scale = params(3);
  int df_max = params(4);
  int num_indices = params(5);

  double epsilon = 0.0001;

  arma::Col<int> num_iter_vec(nrow_tuning_matrix, arma::fill::zeros);
  arma::Col<double> sum_betas(P, arma::fill::zeros);

  arma::Mat<double> beta_matrix(nrow_beta_matrix, ncol_beta_matrix, arma::fill::zeros);
  arma::Mat<double> joint_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<double> temp_b_matrix(P, Q, arma::fill::zeros);
  arma::Mat<int> skip(P, Q, arma::fill::zeros);

  GetRNGstate();

  for (int tuning_index = 0; tuning_index < nrow_tuning_matrix; tuning_index++)
  {
    double lambda1 = tuning_matrix(tuning_index, 0);
    double lambda2 = tuning_matrix(tuning_index, 2);
    double tau2 = tuning_matrix(tuning_index, 3);

    joint_b_matrix.zeros();
    temp_b_matrix.zeros();
    skip.zeros();
    sum_betas.zeros();

    bool converges = true;
    for (int n = 1; n <= num_iter; n++)
    {

      if (num_indices != 0)
      {
        for (int i = 0; i < num_indices; i++)
        {
          int j = index_J(i);

          for (int q = 0; q < Q; q++)
          {
            double bj_bar = summary_betas(j, q);
            if (bj_bar != 0.0)
            {
              double threshold = ((z_scale == 1) * SD_vec(j, q) + (z_scale == 0)) *
                                 (lambda1 + lambda2 / (sum_betas(j) + tau2));

              joint_b_matrix(j, q) = (bj_bar > threshold) * (bj_bar - threshold) +
                                     (bj_bar < -threshold) * (bj_bar + threshold);
            }

            if (summary_betas(j, q) * joint_b_matrix(j, q) < 0)
              perror("Sign inverse error");
          }
        }
      }

      // Finished initialization
      // Now, iterate over all SNPs in LD
      for (int p = 0; p < num_SNP; p++)
      {
        int j = index_matrix(p, 0);

        for (int q = 0; q < Q; q++)
        {
          if (skip(j, q) == 0)
          {
            if (summary_betas(j, q) != 0.0)
            {
              double bj_bar = summary_betas(j, q);

              for (int i = index_matrix(p, 1); i <= index_matrix(p, 2); i++)
              {
                bj_bar -= ld_vec(i) * joint_b_matrix(ld_J(i), q);
              }

              double threshold = ((z_scale == 1) * SD_vec(j, q) + (z_scale == 0)) *
                                 (lambda1 + lambda2 / (sum_betas(j) + tau2));

              if (fabs(bj_bar) > upper_val)
              {
                if (breaking == 1)
                {
                  num_iter_vec(tuning_index) = -1;
                  break;
                }
                else
                {
                  bj_bar = 0.0;
                  skip(j, q) = 1;
                }
              }

              joint_b_matrix(j, q) = (bj_bar > threshold) * (bj_bar - threshold) +
                                     (bj_bar < -threshold) * (bj_bar + threshold);
            }
            else
            {
              joint_b_matrix(j, q) = 0.0;
            }
          }
          else
          {
            joint_b_matrix(j, q) = 0.0;
          }
        }
      }

      // Update the skip matrix
      skip.elem(
              arma::find(
                  arma::abs(joint_b_matrix) > upper_val))
          .ones();

      // Check for divergence
      int df_q = arma::accu(
          arma::abs(joint_b_matrix(arma::span::all, 0)) != 0.0);
      if (df_q > df_max)
        converges = false;

      if (!converges)
      {
        num_iter_vec(tuning_index) = -2;
        break;
      }

      // Check if the beta matrix has converged
      bool found = arma::any(
          arma::find(
              arma::abs(temp_b_matrix - joint_b_matrix) > epsilon));

      if (!found)
      {
        // Determine which entries need to update
        arma::uvec beta_indices = arma::find(joint_b_matrix != 0.0);
        beta_indices.for_each(
            [&](arma::uword idx)
            {
              arma::uvec beta_subscript = arma::ind2sub(arma::size(joint_b_matrix), idx);
              beta_matrix(tuning_index, idx) = joint_b_matrix(beta_subscript(0), beta_subscript(1));
            });

        num_iter_vec(tuning_index) = n;
        break;
      }

      // Update for next iteration if we haven't converged yet
      temp_b_matrix(arma::span::all, arma::span::all) = joint_b_matrix(arma::span::all, arma::span::all);
      sum_betas(arma::span::all) = arma::sum(arma::abs(joint_b_matrix(arma::span::all, arma::span::all)), 1);

      num_iter_vec(tuning_index) = n;
    }
  }
  PutRNGstate();

  return Rcpp::List::create(
      Rcpp::Named("beta_matrix") = beta_matrix,
      Rcpp::Named("num_iter_vec") = num_iter_vec,
      Rcpp::Named("tuning_matrix") = tuning_matrix);
}
